src/Middleware/RouteMiddleware.php

```diff
@@ -0,0 +1,82 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-expressive for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   https://github.com/zendframework/zend-expressive/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Expressive\Middleware;
+
+use Fig\Http\Message\StatusCodeInterface as StatusCode;
+use Interop\Http\ServerMiddleware\DelegateInterface;
+use Interop\Http\ServerMiddleware\MiddlewareInterface as ServerMiddlewareInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\Expressive\Router\RouteResult;
+use Zend\Expressive\Router\RouterInterface;
+
+/**
+ * Default routing middleware.
+ *
+ * Uses the composed router to match against the incoming request.
+ *
+ * When routing failure occurs, if the failure is due to HTTP method, uses
+ * the composed response prototype to generate a 405 response; otherwise,
+ * it delegates to the next middleware.
+ *
+ * If routing succeeds, injects the route result into the request (under the
+ * RouteResult class name), as well as any matched parameters, before
+ * delegating to the next middleware.
+ *
+ * @internal
+ */
+class RouteMiddleware implements ServerMiddlewareInterface
```

Comment:
This was extracted from `Application::routeMiddleware()`. It requires the router instance, as well as a response prototype; the latter is used to create a 405 response when necessary.